community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/IndexingService.java

```diff
@@ -308,9 +314,65 @@ public void accept( Long indexId, IndexProxy proxy )
         samplingController.recoverIndexSamples();
         samplingController.start();
 
+        // So at this point we've started population of indexes that needs to be rebuilt in the background.
+        // Indexes backing uniqueness constraints are normally built within the transaction creating the constraint
+        // and so we shouldn't leave such indexes in a populating state after recovery.
+        // This is why we now go and wait for those indexes to be fully populated.
+        for ( Map.Entry<Long,RebuildingIndexDescriptor> entry : rebuildingDescriptors.entrySet() )
+        {
+            if ( !entry.getValue().getConfiguration().isUnique() )
+            {
+                // It's not a uniqueness constraint, so don't wait for it to be rebuilt
+                continue;
+            }
+
+            IndexProxy proxy;
+            try
+            {
+                proxy = getIndexProxy( entry.getKey().longValue() );
+            }
+            catch ( IndexNotFoundKernelException e )
+            {
+                throw new ThisShouldNotHappenError( "Mattias",
+                        "What? This index was seen during recovery just now, why isn't it available now?" );
+            }
+
+            monitor.awaitingPopulationOfRecoveredIndex( entry.getKey(), entry.getValue().getIndexDescriptor() );
+            awaitOnline( proxy );
+        }
+
         state = State.RUNNING;
     }
 
+    /**
+     * Polls the {@link IndexProxy#getState() state of the index} and waits for it to be either
+     * {@link InternalIndexState#ONLINE}, in which case the wait is over, or {@link InternalIndexState#FAILED},
+     * in which an exception is thrown.
+     */
+    private void awaitOnline( IndexProxy proxy )
+    {
+        while ( true )
+        {
+            switch ( proxy.getState() )
+            {
+            case ONLINE: return;
+            case FAILED: throw new IllegalStateException( "Index entered " + FAILED +
+                    " state while recovery waited for it to be fully populated" );
+            case POPULATING:
+                try
+                {
+                    Thread.sleep( 10 );
+                }
+                catch ( InterruptedException e )
+                {
+                    Thread.currentThread().interrupt();
+                }
+                break;
```

Comment:
This is not very good. If the thread has been interrupted, we will just hot-spin here. There are two or three ways this could be better handled:

1. Set a boolean flag in the catch block, and call `currentThread().interrupt()` in a finally block of this method instead (if the flag is true), in order to allow sleeping to actually take place.
2. Don't catch the exception.
3. *(only if it is safe to do so)* return from this method after catching the exception (and setting the interrupt flag again)